### PR TITLE
ODQA improve

### PR DESCRIPTION
**Pull request by dmitrijeuseew**
ODQA, improve quality and reduce RAM consumption by adding Binary Passage Retrieval and optimizing tfidf index.

### CODE REVIEW

File: deeppavlov/models/torch_bert/torch_transformers_squad.py [hunk 1]
```diff
@@ -76,6 +94,7 @@ def __init__(self,
                  load_before_drop: bool = True,
                  clip_norm: Optional[float] = None,
                  min_learning_rate: float = 1e-06,
+                 psg_cls: bool = False,
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_squad.py:97]: docstring

File: deeppavlov/models/doc_retrieval/bpr.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+from deeppavlov.core.models.component import Component
+from deeppavlov.core.models.serializable import Serializable
+
+
+@register('concat_lists')
+def concat_lists(list_a: List[List[Any]], list_b: List[List[Any]]):
+    list_u = []
+    for element_a, element_b in zip(list_a, list_b):
+        list_u.append(element_a + element_b)
+    return list_u
```
Comment (IgnatovFedor) [inline deeppavlov/models/doc_retrieval/bpr.py:19]: move to separated module. If you keep it here, you will need to add concat_lists to requirements registry

File: deeppavlov/configs/squad/qa_nq_psgcls_bert.json [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      "LOWERCASE": true,
+      "TRANSFORMER": "bert-base-uncased",
+      "ROOT_PATH": "~/.deeppavlov",
+      "DOWNLOADS_PATH": "{ROOT_PATH}/downloads",
+      "MODELS_PATH": "{ROOT_PATH}/models",
+      "MODEL_PATH": "{MODELS_PATH}/passage_reader_classifier_eng"
+    },
+    "download": [
+      {
+        "url": "http://files.deeppavlov.ai/tmp/nq_psgcls_bert.tar.gz",
```
Comment (IgnatovFedor) [inline deeppavlov/configs/squad/qa_nq_psgcls_bert.json:58]: tmp

**Review by IgnatovFedor — CHANGES_REQUESTED** [review]

File: deeppavlov/configs/doc_retrieval/ru_ranker_tfidf_wiki.json [hunk 4]
```diff
@@ -2,58 +2,39 @@
...
         "class_name": "tfidf_ranker",
-        "top_n": 1,
-        "in": [
-          "docs"
-        ],
-        "out": [
-          "tfidf_doc_ids",
-          "tfidf_doc_scores"
-        ],
+        "top_n": 100,
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/configs/doc_retrieval/ru_ranker_tfidf_wiki.json:35]: Yes, more extracted paragraphs increase the probability to contain the correct answer.

File: deeppavlov/configs/doc_retrieval/en_ranker_pop_wiki.json [hunk 5]
```diff
@@ -8,56 +8,39 @@
...
-          "tfidf_doc_scores"
-        ],
+        "top_n": 100,
+        "in": ["docs"],
+        "out": ["tfidf_doc_ids", "tfidf_doc_scores"],
         "vectorizer": "#vectorizer"
       },
       {
         "class_name": "pop_ranker",
-        "pop_dict_path": "{DOWNLOADS_PATH}/odqa/enwiki20180211_popularities.json",
```
Comment (IgnatovFedor) [inline deeppavlov/configs/doc_retrieval/en_ranker_pop_wiki.json:58]: still used in tests/test_configs/doc_retrieval/en_ranker_pop_wiki_test.json
Update test config

File: deeppavlov/configs/doc_retrieval/en_ranker_pop_wiki.json [hunk 6]
```diff
@@ -76,15 +59,15 @@
...
       },
       {
-        "url": "http://files.deeppavlov.ai/deeppavlov_data/en_odqa.tar.gz",
-        "subdir": "{MODELS_PATH}"
+        "url": "http://files.deeppavlov.ai/deeppavlov_data/odqa/enwiki_tfidf_matrix_par_lite.tar.gz",
+        "subdir": "{MODELS_PATH}/odqa"
       },
       {
-        "url": "http://files.deeppavlov.ai/datasets/wikipedia/enwiki20180211_popularities.tar.gz",
+        "url": "http://files.deeppavlov.ai/deeppavlov_data/odqa/enwiki_popularities.tar.gz",
```
Comment (IgnatovFedor) [inline deeppavlov/configs/doc_retrieval/en_ranker_pop_wiki.json:70]: shouldn't file be in datasets section like enwiki20180211_popularities was?

File: deeppavlov/configs/doc_retrieval/en_ranker_pop_wiki.json [hunk 7]
```diff
@@ -76,15 +59,15 @@
     },
     "download": [
       {
-        "url": "http://files.deeppavlov.ai/datasets/wikipedia/enwiki.tar.gz",
-        "subdir": "{DOWNLOADS_PATH}"
+        "url": "http://files.deeppavlov.ai/deeppavlov_data/odqa/enwiki_l100.tar.gz",
```
Comment (IgnatovFedor) [inline deeppavlov/configs/doc_retrieval/en_ranker_pop_wiki.json:62]: shouldn't be in datasets directory like enwiki.tar.gz was?

File: deeppavlov/configs/doc_retrieval/en_ranker_tfidf_wiki.json [hunk 8]
```diff
@@ -70,12 +51,12 @@
     },
     "download": [
       {
-        "url": "http://files.deeppavlov.ai/datasets/wikipedia/enwiki.tar.gz",
-        "subdir": "{DOWNLOADS_PATH}"
+        "url": "http://files.deeppavlov.ai/deeppavlov_data/odqa/enwiki_l100.tar.gz",
```
Comment (IgnatovFedor) [inline deeppavlov/configs/doc_retrieval/en_ranker_tfidf_wiki.json:54]: shouldn't be in datasets section?

File: deeppavlov/configs/doc_retrieval/ru_ranker_tfidf_wiki.json [hunk 9]
```diff
@@ -71,12 +52,12 @@
     },
     "download": [
       {
-        "url": "http://files.deeppavlov.ai/datasets/wikipedia/ruwiki.tar.gz",
-        "subdir": "{DOWNLOADS_PATH}"
+        "url": "http://files.deeppavlov.ai/deeppavlov_data/odqa/ruwiki_par_page_compr.tar.gz",
```
Comment (IgnatovFedor) [inline deeppavlov/configs/doc_retrieval/ru_ranker_tfidf_wiki.json:55]: shouldn't be in datasets directory?

File: deeppavlov/configs/odqa/en_odqa_pop_infer_wiki.json [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+  },
+  "metadata": {
+    "variables": {
+      "ROOT_PATH": "~/.deeppavlov",
+      "DOWNLOADS_PATH": "{ROOT_PATH}/downloads",
+      "MODELS_PATH": "{ROOT_PATH}/models",
+      "CONFIGS_PATH": "{DEEPPAVLOV_PATH}/configs"
+    },
+    "download": [
+    ]
```
Comment (IgnatovFedor) [inline deeppavlov/configs/odqa/en_odqa_pop_infer_wiki.json:57]: empty

File: deeppavlov/configs/odqa/en_odqa_pop_infer_wiki.json [hunk 11]
```diff
@@ -0,0 +1,59 @@
+{
```
Comment (IgnatovFedor) [inline deeppavlov/configs/odqa/en_odqa_pop_infer_wiki.json:1]: Shouldn't this config replace en_odqa_infer_wiki? What benefits do we have from keeping two English odqa models?

**Review by IgnatovFedor — CHANGES_REQUESTED** [review]

File: deeppavlov/configs/doc_retrieval/en_ranker_pop_wiki.json (hunk 7 above)
Comment (IgnatovFedor) [reply deeppavlov/configs/doc_retrieval/en_ranker_pop_wiki.json:62]: is database

File: deeppavlov/configs/doc_retrieval/en_ranker_tfidf_wiki.json (hunk 8 above)
Comment (IgnatovFedor) [reply deeppavlov/configs/doc_retrieval/en_ranker_tfidf_wiki.json:54]: is database

File: deeppavlov/configs/doc_retrieval/en_ranker_tfidf_wiki.json [hunk 12]
```diff
@@ -64,18 +45,18 @@
   },
   "metadata": {
     "variables": {
-      "ROOT_PATH": "~/.deeppavlov",
+      "ROOT_PATH": "/archive/evseev/.deeppavlov",
```
Comment (IgnatovFedor) [inline deeppavlov/configs/doc_retrieval/en_ranker_tfidf_wiki.json:48]: wrong path

**Review by IgnatovFedor — CHANGES_REQUESTED** [review]

**Review by IgnatovFedor — APPROVED** [review]

**Pull request merged**